flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -471,6 +508,7 @@ void jobVertexInFinalState(ExecutionJobVertex ev) {
 						if (current == JobStatus.FAILING) {
 							if (numberOfRetriesLeft > 0 && transitionState(current, JobStatus.RESTARTING)) {
 								numberOfRetriesLeft--;
+								restart();
```

Comment:
This seems wrong here. The restart() call must be asynchronous and should respect the delay.